Python/Product/PythonTools/PythonTools/Environments/AddCondaEnvironmentOperation.cs

```diff
@@ -0,0 +1,281 @@
+                    Title = Strings.CondaStatusCenterCreateTitle.FormatUI(_envName),
+                    DisplayTaskDetails = (t) => { _outputWindow.ShowAndActivate(); }
+                },
+                new TaskProgressData() {
+                    CanBeCanceled = false,
+                    ProgressText = Strings.CondaStatusCenterCreateProgressPreparing,
+                    PercentComplete = null,
+                }
+            );
+
+            var ui = new CondaUI(taskHandler, _outputWindow, _showAndActiveOutputWindow);
+            _statusBar.SetText(Strings.CondaStatusBarCreateStarted.FormatUI(_envName));
+
+            var task = Task.Run(() => CreateCondaEnvironmentAsync(ui, taskHandler, ct));
+            taskHandler.RegisterTask(task);
+            _site.ShowTaskStatusCenter();
+        }
+
+        private async Task CreateCondaEnvironmentAsync(CondaUI ui, ITaskHandler taskHandler, CancellationToken ct) {
+            try {
+                var factory = await CreateFactoryAsync(ui, taskHandler, ct);
+                if (factory != null) {
+                    await _site.GetUIThread().InvokeTask(async () => {
+                        if (_project != null) {
+                            _project.AddInterpreter(factory.Configuration.Id);
+                            if (_setAsCurrent) {
+                                _project.SetInterpreterFactory(factory);
+                            }
+                        }
+
+                        if (_setAsDefault && _options != null) {
+                            _options.DefaultInterpreter = factory;
+                        }
+
+                        if (_viewInEnvWindow) {
+                            await InterpreterListToolWindow.OpenAtAsync(_site, factory);
+                        }
+                    });
+                }
+
+                taskHandler.Progress.Report(new TaskProgressData() {
+                    CanBeCanceled = false,
+                    ProgressText = Strings.CondaStatusCenterCreateProgressCompleted,
+                    PercentComplete = 100,
+                });
+
+                _statusBar.SetText(Strings.CondaStatusBarCreateSucceeded.FormatUI(_envName));
+            } catch (Exception ex) when (!ex.IsCriticalException()) {
+                _statusBar.SetText(Strings.CondaStatusBarCreateFailed.FormatUI(_envName));
+                ui.OnErrorTextReceived(_condaMgr, ex.Message);
+                throw;
+            }
+        }
+
+        private async Task<IPythonInterpreterFactory> CreateFactoryAsync(CondaUI ui, ITaskHandler taskHandler, CancellationToken ct) {
+            var tcs = new TaskCompletionSource<IPythonInterpreterFactory>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            EventHandler interpretersChanged = (object sender, EventArgs e) => {
+                var first = _registry.Interpreters
```

Comment:
Don't we know the `IPythonInterpreterFactoryProvider` for the factory?

Reply:
I would like to get rid of this code, because of the risk of involved in missing the event.

We do know it's `CondaEnvironmentFactoryProvider`, but that factory refreshes after we detect a file system update and fire off a timer to refresh. Meaning, if I query the conda factory provider after creation, it might be too early, it may not have refreshed yet.

I could change it to query/retry/sleep for a certain amount of time instead. In practice the refresh should kick off within a couple of seconds max after the conda mgr function call has returned.

Reply:
Ok, let me take a look on the `CondaEnvironmentFactoryProvider`

Reply:
Ok, it looks like we can add `CondaEnvironmentFactoryProvider.GetFirstFactoryAsync(CancellationToken ct)` that will return first available interpreter that is found in `DiscoverInterpreterFactories`.